// html-validate/dist/cjs/jest-matchers.js
'use strict';

var jestUtils = require('./jest-utils.js');
var core = require('./core.js');
var jestSnapshot = require('jest-snapshot');

function createMatcher$6() {
  function toBeValid(report) {
    if (report.valid) {
      return {
        pass: true,
        message: (
          /* istanbul ignore next */
          () => "Result should not contain error"
        )
      };
    } else {
      const firstError = report.results[0].messages[0];
      return {
        pass: false,
        message: () => `Result should be valid but had error "${firstError.message}"`
      };
    }
  }
  return jestUtils.diverge(toBeValid);
}

function createMatcher$5() {
  function toBeInvalid(report) {
    if (report.valid) {
      return {
        pass: false,
        message: () => "Result should be invalid but had no errors"
      };
    } else {
      return {
        pass: true,
        message: (
          /* istanbul ignore next */
          () => "Result should not contain error"
        )
      };
    }
  }
  return jestUtils.diverge(toBeInvalid);
}

function isMessage(arg) {
  if (!arg || typeof arg !== "object") {
    return false;
  }
  return [
    "ruleId",
    "severity",
    "message",
    "offset",
    "line",
    "column",
    "size",
    "selector",
    "context"
  ].some((key) => key in arg);
}
function isConfig(arg) {
  if (!arg || typeof arg !== "object") {
    return false;
  }
  return ["root", "extends", "elements", "plugin", "transform", "rules"].some((key) => key in arg);
}
function isString(arg) {
  return typeof arg === "string";
}
function getMarkup(src) {
  if (typeof HTMLElement !== "undefined" && src instanceof HTMLElement) {
    return src.outerHTML;
  }
  if (typeof src === "string") {
    return src;
  } else {
    throw new TypeError(`Failed to get markup from "${typeof src}" argument`);
  }
}
function createMatcher$4(expect) {
  function toHTMLValidate(actual, arg0, arg1, arg2) {
    const markup = getMarkup(actual);
    const message = isMessage(arg0) ? arg0 : void 0;
    const config = isConfig(arg0) ? arg0 : isConfig(arg1) ? arg1 : void 0;
    const filename = isString(arg0) ? arg0 : isString(arg1) ? arg1 : arg2;
    return toHTMLValidateImpl.call(this, expect, markup, message, config, filename);
  }
  return jestUtils.diverge(toHTMLValidate);
}
function toHTMLValidateImpl(expect, actual, expectedError, userConfig, filename) {
  const defaultConfig = {
    rules: {
      /* jsdom normalizes style so disabling rule when using this matcher or it
       * gets quite noisy when configured with self-closing */
      "void-style": "off"
    }
  };
  const config = core.deepmerge(defaultConfig, userConfig ?? {});
  const actualFilename = filename ?? this.testPath ?? "inline";
  const syncFn = jestUtils.createSyncFn(core.jestWorkerPath);
  const report = syncFn(actual, actualFilename, config);
  const pass = report.valid;
  const result = report.results[0];
  if (pass) {
    return { pass, message: () => "HTML is valid when an error was expected" };
  } else {
    if (expectedError) {
      const actual2 = result.messages;
      const expected = expect.arrayContaining([expect.objectContaining(expectedError)]);
      const errorPass = this.equals(actual2, expected);
      const diffString = this.utils.diff(expected, actual2, {
        expand: this.expand,
        aAnnotation: "Expected error",
        bAnnotation: "Actual error"
      });
      const hint = this.utils.matcherHint(".not.toHTMLValidate", void 0, void 0, {
        comment: "expected error"
      });
      const expectedErrorMessage = () => [
        hint,
        "",
        "Expected error to be present:",
        this.utils.printExpected(expectedError),
        /* istanbul ignore next */
        diffString ? `
${diffString}` : ""
      ].join("\n");
      return { pass: !errorPass, message: expectedErrorMessage };
    }
    const errors = result.messages.map((message) => `  ${message.message} [${message.ruleId}]`);
    return {
      pass,
      message: () => ["Expected HTML to be valid but had the following errors:", ""].concat(errors).join("\n")
    };
  }
}

function toHaveErrorImpl(context, expect, actual, expected) {
  const flattened = actual.results.flatMap((result) => result.messages);
  const matcher = [expect.objectContaining(expected)];
  const pass = context.equals(flattened, matcher);
  const diffString = context.utils.diff(matcher, flattened, { expand: context.expand });
  const hint = context.utils.matcherHint(".toHaveError");
  const prettyExpected = context.utils.printExpected(matcher);
  const prettyReceived = context.utils.printReceived(flattened);
  const resultMessage = () => {
    return [
      hint,
      "",
      "Expected error to equal:",
      `  ${prettyExpected}`,
      "Received:",
      `  ${prettyReceived}`,
      /* istanbul ignore next */
      diffString ? `
Difference:

${diffString}` : ""
    ].join("\n");
  };
  return { pass, message: resultMessage };
}
function createMatcher$3(expect) {
  function toHaveError(actual, arg1, arg2, arg3) {
    if (typeof arg1 === "string") {
      const expected = {
        ruleId: arg1,
        message: arg2
      };
      if (arg3 !== void 0) {
        expected.context = arg3;
      }
      return toHaveErrorImpl(this, expect, actual, expected);
    } else {
      return toHaveErrorImpl(this, expect, actual, arg1);
    }
  }
  return jestUtils.diverge(toHaveError);
}

function createMatcher$2(expect) {
  function toHaveErrors(report, errors) {
    const flattened = report.results.flatMap((result) => result.messages);
    const matcher = errors.map((entry) => {
      if (Array.isArray(entry)) {
        const [ruleId, message] = entry;
        return expect.objectContaining({ ruleId, message });
      } else {
        return expect.objectContaining(entry);
      }
    });
    const pass = this.equals(flattened, matcher);
    const diffString = this.utils.diff(matcher, flattened, { expand: this.expand });
    const resultMessage = () => this.utils.matcherHint(".toHaveErrors") + `

Expected error to equal:
  ${this.utils.printExpected(matcher)}
Received:
  ${this.utils.printReceived(flattened)}` + /* istanbul ignore next */
    (diffString ? `

Difference:

${diffString}` : "");
    return { pass, message: resultMessage };
  }
  return jestUtils.diverge(toHaveErrors);
}

function getResults(filename, value) {
  if (typeof value === "string") {
    const syncFn = jestUtils.createSyncFn(core.jestWorkerPath);
    const report = syncFn(value, filename, {
      rules: {
        "void-style": "off"
      }
    });
    return report.results.map((it) => {
      return { ...it, filePath: "inline" };
    });
  } else {
    return value.results;
  }
}

function createMatcher$1() {
  function toMatchCodeframe(actual, ...rest) {
    const filename = this.testPath ?? "inline";
    const results = getResults(filename, actual);
    const snapshot = jestUtils.codeframe(results).replaceAll(/\s+$/gm, "");
    return jestSnapshot.toMatchSnapshot.call(this, snapshot, ...rest);
  }
  return jestUtils.diverge(toMatchCodeframe);
}

function toMatchInlineCodeframeImpl(context, actual, ...rest) {
  const filename = context.testPath ?? "inline";
  const results = getResults(filename, actual);
  const snapshot = jestUtils.codeframe(results).replaceAll(/\s+$/gm, "");
  return jestSnapshot.toMatchInlineSnapshot.call(context, snapshot, ...rest);
}
function toMatchInlineCodeframe(actual, ...rest) {
  const context = {
    ...this,
    /* Capture the original stack frames as they are needed by "jest-snapshot"
     * to determine where to write the inline snapshots. When resolving the
     * promise the original stack frames are lost and the snapshot will be
     * written in this files instaed. */
    error: new Error("stacktrace")
  };
  if (jestUtils.isThenable(actual)) {
    return actual.then((resolved) => toMatchInlineCodeframeImpl(context, resolved, ...rest));
  } else {
    return toMatchInlineCodeframeImpl(context, actual, ...rest);
  }
}
function createMatcher() {
  return toMatchInlineCodeframe;
}

exports.createMatcher = createMatcher;
exports.createMatcher$1 = createMatcher$1;
exports.createMatcher$2 = createMatcher$2;
exports.createMatcher$3 = createMatcher$3;
exports.createMatcher$4 = createMatcher$4;
exports.createMatcher$5 = createMatcher$5;
exports.createMatcher$6 = createMatcher$6;
