// html-validate/dist/esm/core-nodejs.js
import fs, { existsSync } from 'node:fs';
import { S as StaticConfigLoader, P as PerformanceTracker, t as transformSource, E as Engine, a as Parser, b as transformSourceSync, c as transformFilename, d as transformFilenameSync, R as Reporter, e as configurationSchema, i as isThenable, U as UserError, C as ConfigLoader, n as normalizeSource, g as ConfigError, h as Config, M as MetaTable, p as parseSeverity, j as ResolvedConfig, f as compatibilityCheckImpl, v as version } from './core.js';
import path from 'node:path';
import fs$1 from 'node:fs/promises';
import { pathToFileURL } from 'node:url';
import { createRequire } from 'node:module';
import kleur from 'kleur';

const defaultFS = {
  readFileSync: fs.readFileSync
};
function isSourceHooks(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  return ["processAttribute", "processElement"].some((key) => key in value);
}
function isConfigData(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  return !["processAttribute", "processElement"].some((key) => key in value);
}
class HtmlValidate {
  configLoader;
  _performanceTracker;
  constructor(arg) {
    const [loader, config] = arg instanceof ConfigLoader ? [arg, void 0] : [void 0, arg];
    this.configLoader = loader ?? new StaticConfigLoader(config);
    this._performanceTracker = null;
  }
  /**
   * Start recording performance data.
   *
   * When active, performance data is accumulated across all subsequent calls to
   * `validateString()`, `validateFile()` and similar methods until
   * {@link HtmlValidate.stopPerformance} is called.
   *
   * @internal
   */
  startPerformance() {
    this._performanceTracker = new PerformanceTracker();
  }
  /**
   * Stop recording performance data and return the result.
   *
   * @internal
   */
  stopPerformance() {
    if (!this._performanceTracker) {
      return { events: [], rules: [], configTime: 0, transformTime: 0, totalTime: 0 };
    }
    const result = this._performanceTracker.getResult();
    this._performanceTracker = null;
    return result;
  }
  /* eslint-enable @typescript-eslint/unified-signatures */
  validateString(str, arg1, arg2, arg3) {
    const filename = typeof arg1 === "string" ? arg1 : "inline";
    const options = isConfigData(arg1) ? arg1 : isConfigData(arg2) ? arg2 : void 0;
    const hooks = isSourceHooks(arg1) ? arg1 : isSourceHooks(arg2) ? arg2 : arg3;
    const source = {
      data: str,
      filename,
      line: 1,
      column: 1,
      offset: 0,
      hooks
    };
    return this.validateSource(source, options);
  }
  /* eslint-enable @typescript-eslint/unified-signatures */
  validateStringSync(str, arg1, arg2, arg3) {
    const filename = typeof arg1 === "string" ? arg1 : "inline";
    const options = isConfigData(arg1) ? arg1 : isConfigData(arg2) ? arg2 : void 0;
    const hooks = isSourceHooks(arg1) ? arg1 : isSourceHooks(arg2) ? arg2 : arg3;
    const source = {
      data: str,
      filename,
      line: 1,
      column: 1,
      offset: 0,
      hooks
    };
    return this.validateSourceSync(source, options);
  }
  /**
   * Parse and validate HTML from [[Source]].
   *
   * @public
   * @param input - Source to parse.
   * @returns Report output.
   */
  async validateSource(input, configOverride) {
    const tracker = this._performanceTracker;
    const source = normalizeSource(input);
    const t0 = performance.now();
    const config = await this.getConfigFor(source.filename, configOverride);
    const resolvers = this.configLoader.getResolvers();
    const t1 = performance.now();
    tracker?.trackConfig(t1 - t0);
    const transformedSource = await transformSource(resolvers, config, source);
    const t2 = performance.now();
    tracker?.trackTransform(t2 - t1);
    const engine = new Engine(config, Parser, { tracker });
    return engine.lint(transformedSource);
  }
  /**
   * Parse and validate HTML from [[Source]].
   *
   * @public
   * @param input - Source to parse.
   * @returns Report output.
   */
  validateSourceSync(input, configOverride) {
    const source = normalizeSource(input);
    const t0 = performance.now();
    const config = this.getConfigForSync(source.filename, configOverride);
    const resolvers = this.configLoader.getResolvers();
    const t1 = performance.now();
    this._performanceTracker?.trackConfig(t1 - t0);
    const transformedSource = transformSourceSync(resolvers, config, source);
    const t2 = performance.now();
    this._performanceTracker?.trackTransform(t2 - t1);
    const engine = new Engine(config, Parser, { tracker: this._performanceTracker });
    return engine.lint(transformedSource);
  }
  /**
   * Parse and validate HTML from file.
   *
   * @public
   * @param filename - Filename to read and parse.
   * @returns Report output.
   */
  async validateFile(filename, fs2 = defaultFS) {
    const tracker = this._performanceTracker;
    const t0 = performance.now();
    const config = await this.getConfigFor(filename);
    const resolvers = this.configLoader.getResolvers();
    const t1 = performance.now();
    tracker?.trackConfig(t1 - t0);
    const source = await transformFilename(resolvers, config, filename, fs2);
    const t2 = performance.now();
    tracker?.trackTransform(t2 - t1);
    const engine = new Engine(config, Parser, { tracker });
    return engine.lint(source);
  }
  /**
   * Parse and validate HTML from file.
   *
   * @public
   * @param filename - Filename to read and parse.
   * @returns Report output.
   */
  validateFileSync(filename, fs2 = defaultFS) {
    const t0 = performance.now();
    const config = this.getConfigForSync(filename);
    const resolvers = this.configLoader.getResolvers();
    const t1 = performance.now();
    this._performanceTracker?.trackConfig(t1 - t0);
    const source = transformFilenameSync(resolvers, config, filename, fs2);
    const t2 = performance.now();
    this._performanceTracker?.trackTransform(t2 - t1);
    const engine = new Engine(config, Parser, { tracker: this._performanceTracker });
    return engine.lint(source);
  }
  /**
   * Parse and validate HTML from multiple files. Result is merged together to a
   * single report.
   *
   * @param filenames - Filenames to read and parse.
   * @returns Report output.
   */
  async validateMultipleFiles(filenames, fs2 = defaultFS) {
    return Reporter.merge(filenames.map((filename) => this.validateFile(filename, fs2)));
  }
  /**
   * Parse and validate HTML from multiple files. Result is merged together to a
   * single report.
   *
   * @param filenames - Filenames to read and parse.
   * @returns Report output.
   */
  validateMultipleFilesSync(filenames, fs2 = defaultFS) {
    return Reporter.merge(filenames.map((filename) => this.validateFileSync(filename, fs2)));
  }
  /**
   * Returns true if the given filename can be validated.
   *
   * A file is considered to be validatable if the extension is `.html` or if a
   * transformer matches the filename.
   *
   * This is mostly useful for tooling to determine whenever to validate the
   * file or not. CLI tools will run on all the given files anyway.
   */
  async canValidate(filename) {
    if (filename.toLowerCase().endsWith(".html")) {
      return true;
    }
    const config = await this.getConfigFor(filename);
    return config.canTransform(filename);
  }
  /**
   * Returns true if the given filename can be validated.
   *
   * A file is considered to be validatable if the extension is `.html` or if a
   * transformer matches the filename.
   *
   * This is mostly useful for tooling to determine whenever to validate the
   * file or not. CLI tools will run on all the given files anyway.
   */
  canValidateSync(filename) {
    if (filename.toLowerCase().endsWith(".html")) {
      return true;
    }
    const config = this.getConfigForSync(filename);
    return config.canTransform(filename);
  }
  /**
   * Tokenize filename and output all tokens.
   *
   * Using CLI this is enabled with `--dump-tokens`. Mostly useful for
   * debugging.
   *
   * @internal
   * @param filename - Filename to tokenize.
   */
  async dumpTokens(filename, fs2 = defaultFS) {
    const config = await this.getConfigFor(filename);
    const resolvers = this.configLoader.getResolvers();
    const source = await transformFilename(resolvers, config, filename, fs2);
    const engine = new Engine(config, Parser, { tracker: null });
    return engine.dumpTokens(source);
  }
  /**
   * Parse filename and output all events.
   *
   * Using CLI this is enabled with `--dump-events`. Mostly useful for
   * debugging.
   *
   * @internal
   * @param filename - Filename to dump events from.
   */
  async dumpEvents(filename, fs2 = defaultFS) {
    const config = await this.getConfigFor(filename);
    const resolvers = this.configLoader.getResolvers();
    const source = await transformFilename(resolvers, config, filename, fs2);
    const engine = new Engine(config, Parser, { tracker: null });
    return engine.dumpEvents(source);
  }
  /**
   * Parse filename and output DOM tree.
   *
   * Using CLI this is enabled with `--dump-tree`. Mostly useful for
   * debugging.
   *
   * @internal
   * @param filename - Filename to dump DOM tree from.
   */
  async dumpTree(filename, fs2 = defaultFS) {
    const config = await this.getConfigFor(filename);
    const resolvers = this.configLoader.getResolvers();
    const source = await transformFilename(resolvers, config, filename, fs2);
    const engine = new Engine(config, Parser, { tracker: null });
    return engine.dumpTree(source);
  }
  /**
   * Transform filename and output source data.
   *
   * Using CLI this is enabled with `--dump-source`. Mostly useful for
   * debugging.
   *
   * @internal
   * @param filename - Filename to dump source from.
   */
  async dumpSource(filename, fs2 = defaultFS) {
    const config = await this.getConfigFor(filename);
    const resolvers = this.configLoader.getResolvers();
    const sources = await transformFilename(resolvers, config, filename, fs2);
    return sources.reduce((result, source) => {
      const line = String(source.line);
      const column = String(source.column);
      const offset = String(source.offset);
      result.push(`Source ${source.filename}@${line}:${column} (offset: ${offset})`);
      if (source.transformedBy) {
        result.push("Transformed by:");
        result = result.concat(source.transformedBy.toReversed().map((name) => ` - ${name}`));
      }
      if (source.hooks && Object.keys(source.hooks).length > 0) {
        result.push("Hooks");
        for (const [key, present] of Object.entries(source.hooks)) {
          if (present) {
            result.push(` - ${key}`);
          }
        }
      }
      result.push("---");
      result = result.concat(source.data.split("\n"));
      result.push("---");
      return result;
    }, []);
  }
  /**
   * Get effective configuration schema.
   */
  getConfigurationSchema() {
    return Promise.resolve(configurationSchema);
  }
  /**
   * Get effective metadata element schema.
   *
   * If a filename is given the configured plugins can extend the
   * schema. Filename must not be an existing file or a filetype normally
   * handled by html-validate but the path will be used when resolving
   * configuration. As a rule-of-thumb, set it to the elements json file.
   */
  async getElementsSchema(filename) {
    const config = await this.getConfigFor(filename ?? "inline");
    const metaTable = config.getMetaTable();
    return metaTable.getJSONSchema();
  }
  /**
   * Get effective metadata element schema.
   *
   * If a filename is given the configured plugins can extend the
   * schema. Filename must not be an existing file or a filetype normally
   * handled by html-validate but the path will be used when resolving
   * configuration. As a rule-of-thumb, set it to the elements json file.
   */
  getElementsSchemaSync(filename) {
    const config = this.getConfigForSync(filename ?? "inline");
    const metaTable = config.getMetaTable();
    return metaTable.getJSONSchema();
  }
  async getContextualDocumentation(message, filenameOrConfig = "inline") {
    const config = typeof filenameOrConfig === "string" ? await this.getConfigFor(filenameOrConfig) : await filenameOrConfig;
    const engine = new Engine(config, Parser, { tracker: null });
    return engine.getRuleDocumentation(message);
  }
  getContextualDocumentationSync(message, filenameOrConfig = "inline") {
    const config = typeof filenameOrConfig === "string" ? this.getConfigForSync(filenameOrConfig) : filenameOrConfig;
    const engine = new Engine(config, Parser, { tracker: null });
    return engine.getRuleDocumentation(message);
  }
  /**
   * Get contextual documentation for the given rule.
   *
   * Typical usage:
   *
   * ```js
   * const report = await htmlvalidate.validateFile("my-file.html");
   * for (const result of report.results){
   *   const config = await htmlvalidate.getConfigFor(result.filePath);
   *   for (const message of result.messages){
   *     const documentation = await htmlvalidate.getRuleDocumentation(message.ruleId, config, message.context);
   *     // do something with documentation
   *   }
   * }
   * ```
   *
   * @internal
   * @deprecated Deprecated since 8.0.0, use [[getContextualDocumentation]] instead.
   * @param ruleId - Rule to get documentation for.
   * @param config - If set it provides more accurate description by using the
   * correct configuration for the file.
   * @param context - If set to `Message.context` some rules can provide
   * contextual details and suggestions.
   */
  async getRuleDocumentation(ruleId, config = null, context = null) {
    const c = config ?? this.getConfigFor("inline");
    const engine = new Engine(await c, Parser, { tracker: null });
    return engine.getRuleDocumentation({ ruleId, context });
  }
  /**
   * Get contextual documentation for the given rule.
   *
   * Typical usage:
   *
   * ```js
   * const report = htmlvalidate.validateFileSync("my-file.html");
   * for (const result of report.results){
   *   const config = htmlvalidate.getConfigForSync(result.filePath);
   *   for (const message of result.messages){
   *     const documentation = htmlvalidate.getRuleDocumentationSync(message.ruleId, config, message.context);
   *     // do something with documentation
   *   }
   * }
   * ```
   *
   * @internal
   * @deprecated Deprecated since 8.0.0, use [[getContextualDocumentationSync]] instead.
   * @param ruleId - Rule to get documentation for.
   * @param config - If set it provides more accurate description by using the
   * correct configuration for the file.
   * @param context - If set to `Message.context` some rules can provide
   * contextual details and suggestions.
   */
  getRuleDocumentationSync(ruleId, config = null, context = null) {
    const c = config ?? this.getConfigForSync("inline");
    const engine = new Engine(c, Parser, { tracker: null });
    return engine.getRuleDocumentation({ ruleId, context });
  }
  /**
   * Create a parser configured for given filename.
   *
   * @internal
   * @param source - Source to use.
   */
  async getParserFor(source) {
    const config = await this.getConfigFor(source.filename);
    return new Parser(config);
  }
  /**
   * Get configuration for given filename.
   *
   * See [[FileSystemConfigLoader]] for details.
   *
   * @public
   * @param filename - Filename to get configuration for.
   * @param configOverride - Configuration to apply last.
   */
  getConfigFor(filename, configOverride) {
    const config = this.configLoader.getConfigFor(filename, configOverride);
    return Promise.resolve(config);
  }
  /**
   * Get configuration for given filename.
   *
   * See [[FileSystemConfigLoader]] for details.
   *
   * @public
   * @param filename - Filename to get configuration for.
   * @param configOverride - Configuration to apply last.
   */
  getConfigForSync(filename, configOverride) {
    const config = this.configLoader.getConfigFor(filename, configOverride);
    if (isThenable(config)) {
      throw new UserError("Cannot use asynchronous config loader with synchronous api");
    }
    return config;
  }
  /**
   * Get current configuration loader.
   *
   * @public
   * @since %version%
   * @returns Current configuration loader.
   */
  /* istanbul ignore next -- not testing setters/getters */
  getConfigLoader() {
    return this.configLoader;
  }
  /**
   * Set configuration loader.
   *
   * @public
   * @since %version%
   * @param loader - New configuration loader to use.
   */
  /* istanbul ignore next -- not testing setters/getters */
  setConfigLoader(loader) {
    this.configLoader = loader;
  }
  /**
   * Flush configuration cache. Clears full cache unless a filename is given.
   *
   * See [[FileSystemConfigLoader]] for details.
   *
   * @public
   * @param filename - If set, only flush cache for given filename.
   */
  flushConfigCache(filename) {
    this.configLoader.flushCache(filename);
  }
}

const legacyRequire = createRequire(import.meta.url);
			const importResolve = (specifier) => {
				return new URL(import.meta.resolve(specifier));
			};

function requireUncached(require, moduleId) {
  const filename = require.resolve(moduleId);
  const m = require.cache[filename];
  if (m?.parent) {
    const { parent } = m;
    for (let i = parent.children.length - 1; i >= 0; i--) {
      if (parent.children[i].id === filename) {
        parent.children.splice(i, 1);
      }
    }
  }
  delete require.cache[filename];
  return require(filename);
}

let cachedRootDir = null;
function determineRootDirImpl(intial, fs2) {
  let current = intial;
  for (; ; ) {
    const search = path.join(current, "package.json");
    if (fs2.existsSync(search)) {
      return current;
    }
    const child = current;
    current = path.dirname(current);
    if (current === child) {
      break;
    }
  }
  return intial;
}
function determineRootDir() {
  cachedRootDir ??= determineRootDirImpl(process.cwd(), fs);
  return cachedRootDir;
}

function expandRelativePath(value, { cwd }) {
  if (typeof value === "string" && value.startsWith(".")) {
    return path.normalize(path.join(cwd, value));
  } else {
    return value;
  }
}

function isRequireError(error) {
  return Boolean(error && typeof error === "object" && "code" in error);
}
function isTransformer$1(value) {
  return typeof value === "function";
}
function cjsResolver(options = {}) {
  const rootDir = options.rootDir ?? determineRootDir();
  function internalRequire(id, { cache }) {
    const moduleName = id.replace("<rootDir>", rootDir);
    try {
      if (cache) {
        return legacyRequire(moduleName);
      } else {
        return requireUncached(legacyRequire, moduleName);
      }
    } catch (err) {
      if (isRequireError(err) && err.code === "MODULE_NOT_FOUND") {
        return null;
      }
      throw err;
    }
  }
  return {
    name: "nodejs-resolver",
    resolveElements(id, options2) {
      return internalRequire(id, options2);
    },
    resolveConfig(id, options2) {
      const configData = internalRequire(id, options2);
      if (!configData) {
        return null;
      }
      const cwd = path.dirname(id);
      const expand = (value) => expandRelativePath(value, { cwd });
      if (Array.isArray(configData.elements)) {
        configData.elements = configData.elements.map(expand);
      }
      if (Array.isArray(configData.extends)) {
        configData.extends = configData.extends.map(expand);
      }
      if (Array.isArray(configData.plugins)) {
        configData.plugins = configData.plugins.map(expand);
      }
      return configData;
    },
    resolvePlugin(id, options2) {
      return internalRequire(id, options2);
    },
    resolveTransformer(id, options2) {
      const mod = internalRequire(id, options2);
      if (!mod) {
        return null;
      }
      if (isTransformer$1(mod)) {
        return mod;
      }
      if (mod.transformer) {
        throw new ConfigError(
          `Module "${id}" is not a valid transformer. This looks like a plugin, did you forget to load the plugin first?`
        );
      }
      throw new ConfigError(`Module "${id}" is not a valid transformer.`);
    }
  };
}

function importFunction(id) {
  return import(id);
}

async function getModuleName(id, { cache, rootDir }) {
  const moduleName = id.replace("<rootDir>", rootDir);
  const url = existsSync(id) ? pathToFileURL(id) : importResolve(moduleName);
  if (url.protocol !== "file:") {
    return url;
  }
  if (cache) {
    return url;
  } else {
    const stat = await fs$1.stat(url);
    url.searchParams.append("mtime", String(stat.mtime.getTime()));
    return url;
  }
}
function isImportError(error) {
  return Boolean(error && typeof error === "object" && "code" in error);
}
async function internalImport(id, rootDir, { cache }) {
  if (id.endsWith(".json")) {
    const content = await fs$1.readFile(id, "utf-8");
    return JSON.parse(content);
  }
  try {
    const url = await getModuleName(id, { cache, rootDir });
    if (url.protocol !== "file:") {
      return null;
    }
    const moduleName = url.toString();
    const { default: defaultImport } = await importFunction(moduleName);
    if (!defaultImport) {
      throw new UserError(`"${id}" does not have a default export`);
    }
    return defaultImport;
  } catch (err) {
    if (isImportError(err) && err.code === "MODULE_NOT_FOUND" && !err.requireStack) {
      return null;
    }
    throw err;
  }
}

function isTransformer(value) {
  return typeof value === "function";
}
function esmResolver(options = {}) {
  const rootDir = options.rootDir ?? determineRootDir();
  return {
    name: "esm-resolver",
    resolveElements(id, options2) {
      return internalImport(id, rootDir, options2);
    },
    async resolveConfig(id, options2) {
      const configData = await internalImport(id, rootDir, options2);
      if (!configData) {
        return null;
      }
      const cwd = path.dirname(id);
      const expand = (value) => expandRelativePath(value, { cwd });
      if (Array.isArray(configData.elements)) {
        configData.elements = configData.elements.map(expand);
      }
      if (Array.isArray(configData.extends)) {
        configData.extends = configData.extends.map(expand);
      }
      if (Array.isArray(configData.plugins)) {
        configData.plugins = configData.plugins.map(expand);
      }
      return configData;
    },
    resolvePlugin(id, options2) {
      return internalImport(id, rootDir, options2);
    },
    async resolveTransformer(id, options2) {
      const mod = await internalImport(id, rootDir, options2);
      if (!mod) {
        return null;
      }
      if (isTransformer(mod)) {
        return mod;
      }
      if (mod.transformer) {
        throw new ConfigError(
          `Module "${id}" is not a valid transformer. This looks like a plugin, did you forget to load the plugin first?`
        );
      }
      throw new ConfigError(`Module "${id}" is not a valid transformer.`);
    }
  };
}

function findConfigurationFiles(fs2, directory) {
  return ["json", "mjs", "cjs", "js"].map((extension) => path.join(directory, `.htmlvalidate.${extension}`)).filter((filePath) => fs2.existsSync(filePath));
}
function normalizeFilename(filename) {
  return filename === "/dev/stdin" ? path.join(process.cwd(), "noop.html") : filename;
}
const defaultResolvers$1 = [esmResolver()];
function hasResolver(value) {
  return Array.isArray(value[0]);
}
class FileSystemConfigLoader extends ConfigLoader {
  cache;
  fs;
  constructor(...args) {
    if (hasResolver(args)) {
      const [resolvers, config, options = {}] = args;
      super(resolvers, config);
      this.fs = /* istanbul ignore next */
      options.fs ?? fs;
    } else {
      const [config, options = {}] = args;
      super(defaultResolvers$1, config);
      this.fs = /* istanbul ignore next */
      options.fs ?? fs;
    }
    this.cache = /* @__PURE__ */ new Map();
  }
  /**
   * Get configuration for given filename.
   *
   * When reading from stdin (`/dev/stdin`), the configuration is searched
   * from the current working directory.
   *
   * @param filename - Filename to get configuration for.
   * @param configOverride - Configuration to merge final result with.
   */
  getConfigFor(filename, configOverride) {
    const normalizedFilename = normalizeFilename(filename);
    const override = this.loadFromObject(configOverride ?? {});
    if (isThenable(override)) {
      return override.then((override2) => {
        return this._resolveAsync(normalizedFilename, override2);
      });
    } else {
      return this._resolveSync1(normalizedFilename, override);
    }
  }
  /**
   * Flush configuration cache.
   *
   * @param filename - If given only the cache for that file is flushed.
   */
  flushCache(filename) {
    if (filename) {
      this.cache.delete(filename);
      const normalized = normalizeFilename(filename);
      if (normalized !== filename) {
        this.cache.delete(normalized);
      }
    } else {
      this.cache.clear();
    }
  }
  /**
   * Load raw configuration from directory traversal.
   *
   * This configuration is not merged with global configuration and may return
   * `null` if no configuration files are found.
   */
  fromFilename(filename) {
    if (filename === "inline") {
      return null;
    }
    const cache = this.cache.get(filename);
    if (cache) {
      return cache;
    }
    let found = false;
    let current = path.resolve(path.dirname(filename));
    let config = this.empty();
    for (; ; ) {
      for (const configFile of findConfigurationFiles(this.fs, current)) {
        const local = this.loadFromFile(configFile);
        if (isThenable(local)) {
          return this.fromFilenameAsync(filename);
        }
        found = true;
        const merged = local.merge(this.resolvers, config);
        if (isThenable(merged)) {
          throw new Error("internal error: async result ended up in sync path");
        }
        config = merged;
      }
      if (config.isRootFound()) {
        break;
      }
      const child = current;
      current = path.dirname(current);
      if (current === child) {
        break;
      }
    }
    if (!found) {
      this.cache.set(filename, null);
      return null;
    }
    this.cache.set(filename, config);
    return config;
  }
  /**
   * Async version of [[fromFilename]].
   *
   * @internal
   */
  async fromFilenameAsync(filename) {
    if (filename === "inline") {
      return null;
    }
    const cache = this.cache.get(filename);
    if (cache) {
      return cache;
    }
    let found = false;
    let current = path.resolve(path.dirname(filename));
    let config = this.empty();
    for (; ; ) {
      for (const configFile of findConfigurationFiles(this.fs, current)) {
        const local = await this.loadFromFile(configFile);
        found = true;
        config = await local.merge(this.resolvers, config);
      }
      if (config.isRootFound()) {
        break;
      }
      const child = current;
      current = path.dirname(current);
      if (current === child) {
        break;
      }
    }
    if (!found) {
      this.cache.set(filename, null);
      return null;
    }
    this.cache.set(filename, config);
    return config;
  }
  _merge(globalConfig, override, config) {
    const merged = config ? config.merge(this.resolvers, override) : globalConfig.merge(this.resolvers, override);
    if (isThenable(merged)) {
      return merged.then((merged2) => {
        return merged2.resolve();
      });
    } else {
      return merged.resolve();
    }
  }
  _resolveSync1(filename, override) {
    if (override.isRootFound()) {
      return override.resolve();
    }
    const globalConfig = this.getGlobalConfig();
    if (isThenable(globalConfig)) {
      return globalConfig.then((globalConfig2) => {
        return this._resolveSync2(filename, override, globalConfig2);
      });
    } else {
      return this._resolveSync2(filename, override, globalConfig);
    }
  }
  _resolveSync2(filename, override, globalConfig) {
    if (globalConfig.isRootFound()) {
      const merged = globalConfig.merge(this.resolvers, override);
      if (isThenable(merged)) {
        return merged.then((merged2) => {
          return merged2.resolve();
        });
      } else {
        return merged.resolve();
      }
    }
    const config = this.fromFilename(filename);
    if (isThenable(config)) {
      return config.then((config2) => {
        return this._merge(globalConfig, override, config2);
      });
    } else {
      return this._merge(globalConfig, override, config);
    }
  }
  async _resolveAsync(filename, override) {
    if (override.isRootFound()) {
      return override.resolve();
    }
    const globalConfig = await this.getGlobalConfig();
    if (globalConfig.isRootFound()) {
      const merged = await globalConfig.merge(this.resolvers, override);
      return merged.resolve();
    }
    const config = await this.fromFilenameAsync(filename);
    return this._merge(globalConfig, override, config);
  }
  /**
   * @internal For testing only
   */
  _getInternalCache() {
    return this.cache;
  }
  defaultConfig() {
    return Config.defaultConfig();
  }
}

function defineFlatConfig(config) {
  return config;
}

const FLAT_CONFIG_FILES = [
  "html-validate.config.mjs",
  "html-validate.config.cjs",
  "html-validate.config.js",
  "html-validate.config.mts",
  "html-validate.config.cts",
  "html-validate.config.ts"
];
function findFlatConfigFile(startDir) {
  let current = path.resolve(startDir);
  const root = path.parse(current).root;
  for (; ; ) {
    for (const filename of FLAT_CONFIG_FILES) {
      const candidate = path.join(current, filename);
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }
    const parent = path.dirname(current);
    if (current === root || parent === current) {
      return null;
    }
    current = parent;
  }
}

async function loadFlatConfigFile(filePath) {
  const url = pathToFileURL(filePath);
  const stat = await fs$1.stat(url);
  url.searchParams.append("mtime", String(stat.mtime.getTime()));
  const module = await import(url.href);
  const value = module.default;
  if (!Array.isArray(value)) {
    throw new TypeError(
      `Flat config file "${filePath}" must have a default export that is an array`
    );
  }
  return value;
}

function mergeFlatConfig(blocks) {
  const merged = {
    elements: [],
    plugins: [],
    transform: {},
    rules: {}
  };
  for (const block of blocks) {
    if (block.elements) {
      merged.elements = [...merged.elements, ...block.elements];
    }
    if (block.plugins) {
      merged.plugins = [...merged.plugins, ...block.plugins];
    }
    if (block.transform) {
      merged.transform = { ...merged.transform, ...block.transform };
    }
    if (block.rules) {
      merged.rules = { ...merged.rules, ...block.rules };
    }
  }
  return merged;
}

function isGlobalIgnore(block) {
  return block.ignores !== void 0 && block.files === void 0 && block.elements === void 0 && block.plugins === void 0 && block.transform === void 0 && block.rules === void 0;
}
function matchesFile(block, rel) {
  if (block.ignores?.some((it) => path.matchesGlob(rel, it))) {
    return false;
  }
  if (!block.files) {
    return true;
  }
  return block.files.some((it) => path.matchesGlob(rel, it.includes("/") ? it : `**/${it}`));
}
function configDataToFlatConfigObject(data) {
  if (!data) {
    return {};
  }
  const elements = data.elements?.filter((it) => typeof it !== "string");
  const plugins = data.plugins?.filter((it) => typeof it !== "string");
  const rules = data.rules;
  return {
    elements,
    plugins,
    rules
  };
}
function buildResolvedConfig(merged, original) {
  const metaTable = new MetaTable();
  for (const element of merged.elements) {
    metaTable.loadFromObject(element);
  }
  metaTable.init();
  const plugins = merged.plugins;
  const rules = /* @__PURE__ */ new Map();
  for (const [ruleId, data] of Object.entries(merged.rules)) {
    let rawSeverity;
    let options;
    if (!Array.isArray(data)) {
      rawSeverity = data;
      options = {};
    } else if (data.length === 1) {
      rawSeverity = data[0];
      options = {};
    } else {
      rawSeverity = data[0];
      options = data[1];
    }
    rules.set(ruleId, [parseSeverity(rawSeverity), options]);
  }
  const transformers = Object.entries(merged.transform).map(
    /* eslint-disable-next-line security/detect-non-literal-regexp -- transform patterns are user-provided regexp strings */
    ([pattern, fn]) => ({ kind: "function", pattern: new RegExp(pattern), function: fn })
  );
  const resolvedData = { metaTable, plugins, rules, transformers };
  return new ResolvedConfig(resolvedData, original);
}
const defaultResolvers = [esmResolver()];
class FlatConfigLoader extends ConfigLoader {
  configFilePath;
  configCache;
  fileCache;
  /**
   * @param configFilePath - Absolute path to the flat config file.
   * @param resolvers - Resolvers to use.
   */
  constructor(configFilePath, resolvers = defaultResolvers) {
    super(resolvers);
    this.configFilePath = configFilePath;
    this.configCache = null;
    this.fileCache = /* @__PURE__ */ new Map();
  }
  /**
   * Search `dir` and its ancestors for a flat config file and, if found,
   * return a new `FlatConfigLoader` pointed at it.  Returns `null` when no
   * config file can be found.
   *
   * @param dir - Directory to start searching from.
   * @param resolvers - Resolvers to use.
   */
  static fromDirectory(dir, resolvers) {
    const configFilePath = findFlatConfigFile(dir);
    if (!configFilePath) {
      return null;
    }
    return new FlatConfigLoader(configFilePath, resolvers);
  }
  /**
   * Get configuration for the given filename.
   *
   * Do note that if `configOverride` is changed between calls the cache must be
   * flushed in-between. This loader caches the final merged result.
   *
   * @param filename - Absolute or relative path to the file being validated.
   * @param configOverride - Optional configuration to merge into the result.
   */
  async getConfigFor(filename, configOverride) {
    const cached = this.fileCache.get(filename);
    if (cached) {
      return cached;
    }
    const { flatConfig, globalIgnores } = await this.loadConfig();
    const configDir = path.dirname(this.configFilePath);
    const rel = path.relative(configDir, path.resolve(filename));
    if (globalIgnores.some((p) => path.matchesGlob(rel, p))) {
      return buildResolvedConfig({ elements: [], plugins: [], transform: {}, rules: {} }, []);
    }
    const overrideBlock = configDataToFlatConfigObject(configOverride);
    const matchingBlocks = flatConfig.filter((block) => matchesFile(block, rel));
    const blocks = [...matchingBlocks, overrideBlock];
    const merged = mergeFlatConfig(blocks);
    const resolved = buildResolvedConfig(merged, blocks);
    this.fileCache.set(filename, resolved);
    return resolved;
  }
  /**
   * Flush the cached flat config, forcing a reload on the next call.
   */
  flushCache(handle) {
    if (handle) {
      this.fileCache.delete(handle);
    } else {
      this.configCache = null;
      this.fileCache = /* @__PURE__ */ new Map();
    }
  }
  /* istanbul ignore next */
  defaultConfig() {
    return this.empty();
  }
  async loadConfig() {
    if (this.configCache) {
      return this.configCache;
    }
    const raw = await loadFlatConfigFile(this.configFilePath);
    const flatConfig = raw.filter((block) => !isGlobalIgnore(block));
    const globalIgnores = raw.filter(isGlobalIgnore).flatMap((block) => block.ignores);
    this.configCache = { flatConfig, globalIgnores };
    return this.configCache;
  }
}

const defaults = {
  silent: false,
  version,
  logger(text) {
    console.error(kleur.red(text));
  }
};
function compatibilityCheck(name, declared, options) {
  return compatibilityCheckImpl(name, declared, {
    ...defaults,
    ...options
  });
}

export { FlatConfigLoader as F, HtmlValidate as H, FileSystemConfigLoader as a, compatibilityCheck as b, cjsResolver as c, defineFlatConfig as d, esmResolver as e, legacyRequire as l };
