// html-validate/dist/esm/index.js
export { a as FileSystemConfigLoader, F as FlatConfigLoader, H as HtmlValidate, c as cjsResolver, b as compatibilityCheck, d as defineFlatConfig, e as esmResolver } from './core-nodejs.js';
export { A as Attribute, h as Config, g as ConfigError, C as ConfigLoader, D as DOMNode, q as DOMTokenList, r as DOMTree, s as DynamicValue, u as EventHandler, H as HtmlElement, w as MetaCopyableProperty, M as MetaTable, N as NestedError, x as Node, a as Parser, P as PerformanceTracker, R as Reporter, j as ResolvedConfig, y as Rule, z as SchemaValidationError, B as Severity, S as StaticConfigLoader, T as TextClassification, F as TextContent, G as TextNode, U as UserError, W as WrappedError, I as ariaNaming, J as classifyNodeText, K as configPresets, L as defineConfig, O as definePlugin, l as formatterFactory, Q as isUserError, V as keywordPatternMatcher, X as ruleExists, Y as sliceLocation, Z as staticResolver, v as version, _ as walk } from './core.js';
export { C as CLI } from './cli.js';
export { d as defineMetadata, m as metadataHelper } from './meta-helper.js';
import 'node:fs';
import 'node:path';
import 'node:fs/promises';
import 'node:url';
import 'node:module';
import 'kleur';
import 'ajv';
import './elements.js';
import './utils/parse-image-candidate-string.js';
import '@sidvind/better-ajv-errors';
import './utils/natural-join.js';
import 'semver';
import '@html-validate/stylish';
import 'node:path/posix';
import 'prompts';
import 'node:util';
