// html-validate/dist/esm/vitest-matchers.js
import { d as diverge, c as createSyncFn, a as codeframe } from './vitest-utils.js';
import { m as deepmerge, a4 as vitestWorkerPath } from './core.js';
import * as vitest from 'vitest';
import { a as FileSystemConfigLoader, H as HtmlValidate } from './core-nodejs.js';

function createMatcher$6() {
  function toBeValid(report) {
    if (report.valid) {
      return {
        pass: true,
        message: (
          /* istanbul ignore next */
          () => "Result should not contain error"
        )
      };
    } else {
      const firstError = report.results[0].messages[0];
      return {
        pass: false,
        message: () => `Result should be valid but had error "${firstError.message}"`
      };
    }
  }
  return diverge(toBeValid);
}

function createMatcher$5() {
  function toBeInvalid(report) {
    if (report.valid) {
      return {
        pass: false,
        message: () => "Result should be invalid but had no errors"
      };
    } else {
      return {
        pass: true,
        message: (
          /* istanbul ignore next */
          () => "Result should not contain error"
        )
      };
    }
  }
  return diverge(toBeInvalid);
}

function isMessage(arg) {
  if (!arg || typeof arg !== "object") {
    return false;
  }
  return [
    "ruleId",
    "severity",
    "message",
    "offset",
    "line",
    "column",
    "size",
    "selector",
    "context"
  ].some((key) => key in arg);
}
function isConfig(arg) {
  if (!arg || typeof arg !== "object") {
    return false;
  }
  return ["root", "extends", "elements", "plugin", "transform", "rules"].some((key) => key in arg);
}
function isString(arg) {
  return typeof arg === "string";
}
function getMarkup(src) {
  if (typeof HTMLElement !== "undefined" && src instanceof HTMLElement) {
    return src.outerHTML;
  }
  if (typeof src === "string") {
    return src;
  } else {
    throw new TypeError(`Failed to get markup from "${typeof src}" argument`);
  }
}
function createMatcher$4(expect) {
  function toHTMLValidate(actual, arg0, arg1, arg2) {
    const markup = getMarkup(actual);
    const message = isMessage(arg0) ? arg0 : void 0;
    const config = isConfig(arg0) ? arg0 : isConfig(arg1) ? arg1 : void 0;
    const filename = isString(arg0) ? arg0 : isString(arg1) ? arg1 : arg2;
    return toHTMLValidateImpl.call(this, expect, markup, message, config, filename);
  }
  return diverge(toHTMLValidate);
}
function toHTMLValidateImpl(expect, actual, expectedError, userConfig, filename) {
  const defaultConfig = {
    rules: {
      /* jsdom normalizes style so disabling rule when using this matcher or it
       * gets quite noisy when configured with self-closing */
      "void-style": "off"
    }
  };
  const config = deepmerge(defaultConfig, userConfig ?? {});
  const actualFilename = filename ?? this.testPath ?? "inline";
  const syncFn = createSyncFn(vitestWorkerPath);
  const report = syncFn(actual, actualFilename, config);
  const pass = report.valid;
  const result = report.results[0];
  if (pass) {
    return { pass, message: () => "HTML is valid when an error was expected" };
  } else {
    if (expectedError) {
      const actual2 = result.messages;
      const expected = expect.arrayContaining([expect.objectContaining(expectedError)]);
      const errorPass = this.equals(actual2, expected);
      const diffString = this.utils.diff(expected, actual2, {
        aAnnotation: "Expected error",
        bAnnotation: "Actual error"
      });
      const hint = this.utils.matcherHint(".not.toHTMLValidate", void 0, void 0, {
        comment: "expected error"
      });
      const expectedErrorMessage = () => [
        hint,
        "",
        "Expected error to be present:",
        this.utils.printExpected(expectedError),
        /* istanbul ignore next */
        diffString ? `
${diffString}` : ""
      ].join("\n");
      return { pass: !errorPass, message: expectedErrorMessage, actual: actual2, expected };
    }
    const errors = result.messages.map((message) => `  ${message.message} [${message.ruleId}]`);
    return {
      pass,
      message: () => ["Expected HTML to be valid but had the following errors:", ""].concat(errors).join("\n")
    };
  }
}

function toHaveErrorImpl(context, expect, actual, expected) {
  const flattened = actual.results.flatMap((result) => result.messages);
  const matcher = [expect.objectContaining(expected)];
  const pass = context.equals(flattened, matcher);
  const diffString = context.utils.diff(matcher, flattened);
  const hint = context.utils.matcherHint(".toHaveError");
  const prettyExpected = context.utils.printExpected(matcher);
  const prettyReceived = context.utils.printReceived(flattened);
  const resultMessage = () => {
    return [
      hint,
      "",
      "Expected error to equal:",
      `  ${prettyExpected}`,
      "Received:",
      `  ${prettyReceived}`,
      /* istanbul ignore next */
      diffString ? `
Difference:

${diffString}` : ""
    ].join("\n");
  };
  return { pass, message: resultMessage, actual: flattened, expected: matcher };
}
function createMatcher$3(expect) {
  function toHaveError(actual, arg1, arg2, arg3) {
    if (typeof arg1 === "string") {
      const expected = {
        ruleId: arg1,
        message: arg2
      };
      if (arg3 !== void 0) {
        expected.context = arg3;
      }
      return toHaveErrorImpl(this, expect, actual, expected);
    } else {
      return toHaveErrorImpl(this, expect, actual, arg1);
    }
  }
  return diverge(toHaveError);
}

function createMatcher$2(expect) {
  function toHaveErrors(report, errors) {
    const flattened = report.results.flatMap((result) => result.messages);
    const matcher = errors.map((entry) => {
      if (Array.isArray(entry)) {
        const [ruleId, message] = entry;
        return expect.objectContaining({ ruleId, message });
      } else {
        return expect.objectContaining(entry);
      }
    });
    const pass = this.equals(flattened, matcher);
    const diffString = this.utils.diff(matcher, flattened);
    const resultMessage = () => this.utils.matcherHint(".toHaveErrors") + `

Expected error to equal:
  ${this.utils.printExpected(matcher)}
Received:
  ${this.utils.printReceived(flattened)}` + /* istanbul ignore next */
    (diffString ? `

Difference:

${diffString}` : "");
    return { pass, message: resultMessage };
  }
  return diverge(toHaveErrors);
}

function createMatcher$1() {
  const loader = new FileSystemConfigLoader({
    extends: ["html-validate:recommended"]
  });
  const htmlvalidate = new HtmlValidate(loader);
  async function toMatchCodeframe(received, hint) {
    if (!vitest.Snapshots) {
      throw new Error("toMatchCodeframe() requires vitest 4.1.3 or later. Please upgrade vitest.");
    }
    const resolved = await received;
    let report;
    if (typeof resolved === "string") {
      const filename = this.testPath ?? "inline";
      report = await htmlvalidate.validateString(resolved, filename, {
        rules: {
          "void-style": "off"
        }
      });
    } else {
      report = resolved;
    }
    const snapshot = codeframe(report.results).replaceAll(/\s+$/gm, "");
    return vitest.Snapshots.toMatchSnapshot.call(this, snapshot, hint);
  }
  return toMatchCodeframe;
}

function createMatcher() {
  const loader = new FileSystemConfigLoader({
    extends: ["html-validate:recommended"]
  });
  const htmlvalidate = new HtmlValidate(loader);
  async function toMatchInlineCodeframe(received, inlineSnapshot) {
    if (!vitest.Snapshots) {
      throw new Error(
        "toMatchInlineCodeframe() requires vitest 4.1.3 or later. Please upgrade vitest."
      );
    }
    vitest.chai.util.flag(this.assertion, "error", new Error("stacktrace"));
    const resolved = await received;
    let report;
    if (typeof resolved === "string") {
      const filename = this.testPath ?? "inline";
      report = await htmlvalidate.validateString(resolved, filename, {
        rules: {
          "void-style": "off"
        }
      });
    } else {
      report = resolved;
    }
    const snapshot = codeframe(report.results).replaceAll(/\s+$/gm, "");
    return vitest.Snapshots.toMatchInlineSnapshot.call(this, snapshot, inlineSnapshot);
  }
  return toMatchInlineCodeframe;
}

export { createMatcher$1 as a, createMatcher$2 as b, createMatcher as c, createMatcher$3 as d, createMatcher$4 as e, createMatcher$5 as f, createMatcher$6 as g };
